pyomo/gdp/plugins/cuttingplane.py

```diff
@@ -15,89 +15,688 @@
+                              solution in order to be added to the bigM model
+    zero_tolerance : Tolerance at which a float will be considered 0 when
+                     using Fourier-Motzkin elimination to create cuts.
+    do_integer_arithmetic : Whether or not to require Fourier-Motzkin elimination
+                            to do integer arithmetic. Only possible when all
+                            data is integer.
+
+    By default, the callbacks will be set such that the algorithm performed is
+    as presented in [1], but with an additional post-processing procedure to
+    reduce numerical error, which calculates the maximum violation of the cut 
+    subject to the relaxed hull constraints, and then pads the constraint by 
+    this violation plus an additional user-specified tolerance.
+
+    In addition, the create_cuts_fme function provides an (exponential time)
+    method of generating cuts which reduces numerical error (and can eliminate 
+    it if all data is integer). It collects the hull constraints which are 
+    tight at the solution of the separation problem. It creates a cut in the 
+    extended space perpendicular to  a composite normal vector created by 
+    summing the directions normal to these constraints. It then performs 
+    fourier-motzkin elimination on the collection of constraints and the cut
+    to project out the disaggregated variables. The resulting constraint which
+    is most violated by the relaxed bigM solution is then returned.
+
+    References
+    ----------
+        [1] Sawaya, N. W., Grossmann, I. E. (2005). A cutting plane method for 
+        solving linear generalized disjunctive programming problems. Computers
+        and Chemical Engineering, 29, 1891-1913 
+    """
+
+    CONFIG = ConfigBlock("gdp.cuttingplane")
+    CONFIG.declare('solver', ConfigValue(
+        default='ipopt',
+        domain=str,
+        description="""Solver to use for relaxed BigM problem and the separation
+        problem""",
+        doc="""
+        This specifies the solver which will be used to solve LP relaxation
+        of the BigM problem and the separation problem. Note that this solver
+        must be able to handle a quadratic objective because of the separation
+        problem.
+        """
+    ))
+    CONFIG.declare('minimum_improvement_threshold', ConfigValue(
+        default=0.01,
+        domain=PositiveFloat,
+        description="Threshold value for difference in relaxed bigM problem "
+        "objectives used to decide when to stop adding cuts",
+        doc="""
+        If the difference between the objectives in two consecutive iterations is
+        less than this value, the algorithm terminates without adding the cut
+        generated in the last iteration.  
+        """
+    ))
+    CONFIG.declare('separation_objective_threhold', ConfigValue(
+        default=0.01,
+        domain=PositiveFloat,
```

Comment:
I think not because you would be asking for a solver to get an exactly 0 objective. Which seems scary since you'd have an infinite loop if it didn't work out.